### ISSUE TITLE
Taking rendered window screenshots fails with vo=vaapi

### ISSUE THREAD

**Issue by DragoonAethis** [mpv-player/mpv#6869]
Trying to take a "rendered" screenshot (Ctrl+s) with VAAPI video output fails.

### mpv version and platform

Reproducible on both mpv 0.29.1 and current master (cd7bcb9) on Arch Linux. VAAPI uses the libva-intel-driver 2.3.0-2 (HD 530, low-end desktop Skylake graphics) and libva 2.5.0-1.

### Reproduction steps

With the default config, pseudo GUI:
- `mpv test.mkv`, press `Ctrl-s`: Works
- `mpv test.mkv -vo=vaapi`, press `Ctrl-s`: **Taking screenshot failed.**

# Files

- Log: https://0x0.st/zVGx.log
- Sample: https://0x0.st/zVGY.mkv (starts loudly, `-mute` before use :)

**Comment by CounterPillow** [comment]
don't use `vo=vaapi`.

**Comment by DragoonAethis** [comment]
I know, but my potato GPU can't do high bitrate 1080p on `vo=gpu` ;-;

**Comment by DragoonAethis** [comment]
Realistically, I know it sucks and that fixing this may be a pain, but it could be at least noted in the error message in some way, like "Taking rendered screenshots doesn't work with the current video output (vaapi)." instead of just failing for no apparent reason.

**Comment by CounterPillow** [comment]
bitrate has nothing to do with vo=gpu performance. If you need hardware decoding, use `hwdec=vaapi`. I also don't see why you'd want to take a rendered screenshot with a low-quality VO.

**Comment by CounterPillow** [comment]
Seeing as I'm on an HD 620 and can do gpu-hq with `scale=ewa_lanczossharp`, I'd like to see a log of you playing a file with the gpu VO to make sure you're not doing something weird with your drivers.

**Comment by DragoonAethis** [comment]
> I also don't see why you'd want to take a rendered screenshot with a low-quality VO.

For comparison purposes between `profile=low-latency` and `profile=gpu-hq` with [extra shaders](https://github.com/bloc97/Anime4K).

Regarding the performance issues... hmm, looks like you're right? I had this config running for quite a few years and it looks like `vo=gpu` with `profile=default` now works just fine even on 4K videos. Previously I had issues with something like [this](https://0x0.st/zVG7.mp4) that would cause a lot of dropped frames and the desktop itself lagging as well (KDE 5 w/ OpenGL 3.1 compositor enabled), but apparently whatever was the problem is no longer a thing. Thanks! (It still doesn't fix the bug, though :)

**Comment by CounterPillow** [comment]
profile=gpu-hq has absolutely no effect on the vaapi VO, and shaders don't get loaded either since it's not a shader based renderer. At best, the vaapi VO is equivalent to the dumb-mode GPU renderer (bilinear scaling, no dithering, etc.) At worst it's worse by scuffing up the colours or doing other weird stuff. So if you want to compare Anime4K to not using Anime4K, you should compare it to the gpu vo as a baseline anyway.

>(It still doesn't fix the bug, though :)

The vaapi VO is pretty much a black box and I'm not sure if there even is an API to get rendered data back out of it.

**Comment by DragoonAethis** [comment]
It's completely off topic from this bug, but here's what I tested:

- Config w/ `vo=vaapi`
- `mpv test.mkv` w/ no shaders, results as usual
- `...` w/ shaders, nothing changed...?
- `... -vo=gpu` w/ shaders, much better now
` `... -vo=gpu` w/ no shaders, okay, that shader does make a difference
- `... -vo=gpu-hq`, oops, that's not a thing
` `... -vo=gpu -profile=gpu-hq` w/ shaders - a bit better, but not that much

At which point I wanted to screenshot a few of these for some folks on the chat, starting from the original `vo=vaapi`, `s`... but it saves a 480p screenshot, can I get the rendered output? Yes, `Ctrl+s`, but it fails. And with `vo=vaapi` only, `vo=gpu` works.

**Comment by DragoonAethis** [comment]
>The vaapi VO is pretty much a black box and I'm not sure if there even is an API to get rendered data back out of it.

It might be with [vaDeriveImage](https://intel.github.io/libva/group__api__core.html#gaaddbbd4c6ea5bcfe928adc991cc5faab), maybe?

**Comment by CounterPillow** [comment]
I'm just saying it's a pointless feature, but I actually did some research just for the hell of it:

What you'd need to do is implement `VOCTRL_SCREENSHOT` in the vo_vaapi.c control function, but supporting [all the flags for `VOCTRL_SCREENSHOT`](https://github.com/mpv-player/mpv/blob/199aabddcc0105b504cc8e0cb240bc3c89288c1a/video/out/vo.h#L186-L189) seems like a pain and would probably need dicking around with disassociating subpictures conditionally and then reading the data back, and also creating a different surface if we want it unscaled if I understand the API correctly. The `vaDeriveImage` call also isn't supported by all VA-API implementations, so you'd need to add fallback code to make sure the whole thing doesn't crash and burn if someone on a core2duo thinkpad hits ctrl+s.

So in conclusion, you'd need to implement some duplicate code that doesn't quite work right for something that's probably going to look just like a normal `s` or `S` screenshot. Seems like pointless work to support a function that mostly was added to the gpu VO so people can take high quality screenshots using mpv's renderer instead of ffmpeg's swscale. Pretty sure no other VO supports it (vdpau supports `VOCTRL_SCREENSHOT_WIN` though.)

**Comment by DragoonAethis** [comment]
Fair enough, sounds like a pain. Closing for now then.
